### DISCUSSION TITLE
How do I selectively perform certain responses tasks or targets based on item properties

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [dotnet/msbuild#7524] [Q&A] [answered]
Let's say we have:

```xml
<ItemGroup>
  <MySpec Include="Alpha" Type="Module" ... />
  <MySpec Include="Bravo" Type="Feature" ... />
  <MySpec Include="Charlie" Type="Feature" ... />
  <MySpec Include="Delta" Type="Theme" ... />
  <MySpec Include="Echo" Type="Theme" ... />
  <MySpec Include="Foxtrot" Type="SomethingElse" ... />
</ItemGroup>
```

My goal there, if possible, is to filter on all `Module`, `Feature`, `Theme`, then anything else with the `Type` as such. Is it possible?

And while it would be _'simple enough'_ to add specialized item lists, I would like to avoid that if at all possible, and simply support a single variant style shape.

I think that the solution must probably involve some sort of copying from one source items list to another, using a condition, predicate, or what have you, but the tasks, inputs, outputs, etc, are a bit unfamiliar to me.

**Comment by mwpowellhtx** [comment]
For instance, [`FindInList`](https://docs.microsoft.com/en-us/visualstudio/msbuild/findinlist-task) seems like a possible application here, but would be fantastic to have some examples. i.e. given the source list items, find what/how, directed to what targets, and so forth.

**Comment by jrdodds** [comment, marked as answer]
It sounds like you are asking about how to do [batching](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching?view=vs-2022).

Here is an example of [task batching](https://docs.microsoft.com/en-us/visualstudio/msbuild/item-metadata-in-task-batching?view=vs-2022) using your example data:
```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <ItemGroup>
        <MySpec Include="Alpha" Type="Module" />
        <MySpec Include="Bravo" Type="Feature" />
        <MySpec Include="Charlie" Type="Feature" />
        <MySpec Include="Delta" Type="Theme" />
        <MySpec Include="Echo">
            <Type>Theme</Type>
        </MySpec>
        <MySpec Include="Foxtrot" Type="SomethingElse" />
    </ItemGroup>

    <Target Name="Test">
        <Message Text="Features:     @(MySpec, ', ')" Condition="%(Type) == 'Feature'"/>
        <Message Text="Modules:      @(MySpec, ', ')" Condition="%(Type) == 'Module'"/>
        <Message Text="Misc:         @(MySpec, ', ')" Condition="%(Type) == 'Misc'"/>
        <Message Text="Themes+Other: @(MySpec, ', ')" Condition="%(Type) != 'Feature' and %(Type) != 'Module'"/>

        <ItemGroup>
            <SubSet Include="@(MySpec)" Condition="%(MySpec.Type) != 'Feature' and %(MySpec.Type) != 'Module'" />
        </ItemGroup>
        <Message Text="Subset:       @(Subset, ', ')" />
    </Target>
</Project>
```

The output will be:
```
         Features:     Bravo, Charlie
         Modules:      Alpha
         Themes+Other: Delta, Echo
         Themes+Other: Foxtrot
         Subset:       Delta, Echo, Foxtrot
```

MSBuild is declarative. There are no loops. Don't think of Targets and Tasks as function calls.

In the example, the first four `Message` tasks use the `Type` metadata. Because the metadata is referenced, the `MySpec` item group collection is split into batches based on `Type` and the task is performed for each batch.

The conditions on each `Message` task are where the `Type` metadata is referenced which causes the batching. But the conditions also determine if the task is performed for a given batch. Note that the `Message` task for 'Misc' is not performed because nothing matches. But also note that the 'Themes+Other' task is run twice, once for each of two different batches of `Type`.

In many cases it is fine to perform a task multiple times with different batches. For the times when that is not acceptable, batching can be used to create a new item group collection. The `SubSet` item group is an example.

**Reply by mwpowellhtx** [reply]
Great response, thanks! I did do some background reading, I get it; they are not functions, unless they _'are'_ i.e. taking inputs returning outputs, but otherwise, yes, totally get it. Did not know I could do the batch-wise condition, this is awesome, just what I was looking for.
